crates/goose/src/systems/non_developer.rs

```diff
@@ -0,0 +1,971 @@
+            .await?;
+
+        // Register as a resource
+        let uri = Url::from_file_path(&cache_path)
+            .map_err(|_| AgentError::ExecutionError("Invalid cache path".into()))?
+            .to_string();
+
+        let resource = Resource::new(
+            uri.clone(),
+            Some("json".to_string()),
+            Some(cache_path.to_string_lossy().into_owned()),
+        )
+        .map_err(|e| AgentError::ExecutionError(e.to_string()))?;
+
+        self.active_resources.lock().unwrap().insert(uri, resource);
+
+        Ok(vec![Content::text(format!(
+            "Search results saved to: {}",
+            cache_path.display()
+        ))])
+    }
+
+    async fn web_scrape(&self, params: Value) -> AgentResult<Vec<Content>> {
+        let url = params
+            .get("url")
+            .and_then(|v| v.as_str())
+            .ok_or_else(|| AgentError::InvalidParameters("Missing 'url' parameter".into()))?;
+
+        let save_as = params
+            .get("save_as")
+            .and_then(|v| v.as_str())
+            .unwrap_or("text");
+
+        // Fetch the content
+        let response = self
+            .http_client
+            .get(url)
+            .send()
+            .await
+            .map_err(|e| AgentError::ExecutionError(format!("Failed to fetch URL: {}", e)))?;
+
+        let status = response.status();
+        if !status.is_success() {
+            return Err(AgentError::ExecutionError(format!(
+                "HTTP request failed with status: {}",
+                status
+            )));
+        }
+
+        // Process based on save_as parameter
+        let (content, extension) = match save_as {
+            "text" => {
+                let text = response.text().await.map_err(|e| {
+                    AgentError::ExecutionError(format!("Failed to get text: {}", e))
+                })?;
+                (text.into_bytes(), "txt")
+            }
+            "json" => {
+                let text = response.text().await.map_err(|e| {
+                    AgentError::ExecutionError(format!("Failed to get text: {}", e))
+                })?;
+                // Verify it's valid JSON
+                serde_json::from_str::<Value>(&text).map_err(|e| {
+                    AgentError::ExecutionError(format!("Invalid JSON response: {}", e))
+                })?;
+                (text.into_bytes(), "json")
+            }
+            "binary" => {
+                let bytes = response.bytes().await.map_err(|e| {
+                    AgentError::ExecutionError(format!("Failed to get bytes: {}", e))
+                })?;
+                (bytes.to_vec(), "bin")
+            }
+            _ => unreachable!(), // Prevented by enum in tool definition
+        };
+
+        // Save to cache
+        let cache_path = self.save_to_cache(&content, "web", extension).await?;
+
+        // Register as a resource
+        let uri = Url::from_file_path(&cache_path)
```

Comment:
maybe add a register_as_a_resource helper function since you use this block throughout

Reply:
oh good idea